engines/ic3base.h

```diff
@@ -0,0 +1,465 @@
+** \verbatim
+** Top contributors (to current version):
+**   Makai Mann, Ahmed Irfan
+** This file is part of the pono project.
+** Copyright (c) 2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief Abstract base class implementation of IC3 parameterized by
+**        the unit used in frames, pre-image computation, and inductive
+**        and predecessor generalization techniques.
+**
+**        To create a particular IC3 instantiation, you must implement the
+*following:
+**           - implement get_ic3_formula and give it semantics to produce the
+**             corresponding IC3Formula for your flavor of IC3
+**             (assumes solver_'s state is SAT from a failed rel_ind_check)
+**             also need to be able to give model (as formulas) for input values
+**             and next-state variable values if inputs/nexts are non-null,
+*respectively
+**           - implement inductive_generalization
+**           - implement generalize_predecessor
+**             this one is special because it is called with solver_context == 1
+**             and assuming the state is SAT so you can query with
+**             solver_->get_value(Term)
+**             However, this also means you can't use the solver_ because
+**             it's polluted with other assertions. All calls are expected
+**             to be through a separate solver (e.g. through the reducer_)
+**           - implement check_ts which just checks if there are any
+**             theories / syntax used in the transition system which
+**             is not supported by this instantiation
+**          [OPTIONAL]
+**           - implement all the ic3_formula_* functions for creating and
+**             manipulating an IC3Formula if the defaults are not right
+**           - implement abstract() and refine() if this is a CEGAR
+**             flavor of IC3
+**
+**        Important Notes:
+**           - be sure to use [push/pop]_solver_context instead of using
+**             the solver interface directly so that the assertions on
+**             the solver context (tracked externally) are correct
+**
+**/
+#pragma once
+
+#include "engines/prover.h"
+#include "smt-switch/utils.h"
+
+namespace pono {
+
+struct IC3Formula
+{
+  // nullary constructor
+  IC3Formula() : term(nullptr) {}
+  IC3Formula(const smt::Term & t, const smt::TermVec & c, bool n)
+      : term(t), children(c), disjunction(n)
+  {
+  }
+
+  IC3Formula(const IC3Formula & other)
+      : term(other.term),
+        children(other.children),
+        disjunction(other.disjunction)
+  {
+  }
+
+  virtual ~IC3Formula() {}
+
+  /** Returns true iff this IC3Formula has not been initialized */
+  bool is_null() const { return (term == nullptr); };
+
+  bool is_disjunction() const { return disjunction; };
```

Comment:
To avoid over-generalizing, decided to just assume that an `IC3Formula` is always a disjunction or conjunction. The advantage of this is that it allows more work to be done in the base class instead of relying on the derived classes to do it. For example, this allows for using the unsat core for a first round of generalization if you know the structure. This could be changed later.